### ISSUE TITLE
Unintended UV Behavior of Draw_Sprite_Part()

### ISSUE THREAD

**Issue by mistletoe** [YoYoGames/GameMaker-Bugs#2582]
### Description

draw_sprite_part() and all of the functions that rely on it (draw_sprite_part_ext(), draw_sprite_general()) behave unintuitively regarding the treatment of UV coordinates. A relatively simple fix to the existing engine code would correct this problem.

What happens is pretty obvious: if draw_sprite_part's arguments permit the draw call to exceed the UV boundaries of the Sprite, then the UVs are shifted outside the boundaries. 

For example, this call to draw_sprite_part will cause UV overflow:  

_draw_sprite_part(myTestSprite,0,-15,-15,sprite_height+15,sprite_height+15,x+200,y+200);_

Due to how this is implemented, it doesn't really behave like how it's described in the manual, in practice. For example, if we want to do a simple health-bar using a ratio to determine the length of the bar, it's fine, so long as the sprite_draw_part is cutting off the bottom part of the Sprite:

draw_sprite_part(sprite_index,0,0,0,sprite_width,sprite_height  * healthRatio,x-sprite_width*0.5,y-sprite_height*0.5);

This operates as expected in the documentation. However, if we change the code like this, to make the sprite get cut off from the top, you'll immediately see the problem:

draw_sprite_part(sprite_index,0,0,0+(1/healthRatio),sprite_width,sprite_height,x-sprite_width*0.5,y-sprite_height*0.5);

What happens here is that the V value being output for both the top-left and bottom-right corners of the quad are wrong; the Vs are getting displaced "downwards" for both the top-left and bottom-right coordinates. According to the example in the GML manual, this should work, but it doesn't; putting positive values into the first two XY coordinates causes UV over/underflow.

To correct this problem is possible in GML:

var sprRatioHeight = sprite_height*healthRatio;
var dist = sprite_height-(sprRatioHeight);
draw_sprite_part(sprite_index,0,0,0+dist,sprite_width,sprRatioHeight,x-sprite_width*0.5,y-sprite_height*0.5+dist);

Now we have a simple health bar that grows smaller from the top downwards. But figuring out the ratios, understanding what's going on in terms of UVs, etc. is a little advanced for newbies. 

To fix this stuff, the code for draw_sprite_part needs to presume:

1. The first XY, if it contains values less than 0, are set to 0.
3. Calls to the second XY more than sprite_width or sprite_height (respectively) should be set to those values.
4. The last pair of XY values should just be moving around the quad's location in the world. Not moving around the displaced UVs.

Now, I understand that some clever people may have figured out that they can use this as a way to scroll UVs, and they may be annoyed if this unintended functionality is removed, which is fair (it's only an exploit if it's called a bug, after all). 

But they'll be fine. They can easily reproduce scrolling behaviors with a shader... and that will also give them other advantages, such as sub-pixel accuracy. 

Since it's best to provide a solution rather than just a problem, here's a simple scrolling shader's vertex instructions moving the U over time.

//FIXED GMS INPUTS
attribute vec3 in_Position;                  // (x,y,z)
attribute vec4 in_Colour;                    // (r,g,b,a)
attribute vec2 in_TextureCoord;              // (u,v)

//HELPFUL CONSTANTS
float sizeOfMyTextureAtlas = 2048.0;
float sizeOfMyBorder= 2.0;
float trimDisplacement = 1/sizeOfMyTextureAtlas * sizeOfMyBorder;

//INPUT VAR
uniform float iTime;//Time value, from GML
uniform vec4 UVs;//The UV values of the Sprite's top-left and bottom-right, respectively.

//OUTPUT
varying vec4 v_Color;
varying vec2 v_TexCoord;

void main()
{
	 //Standard world-space position of vertices; the quad isn't moving, just the U.
	 vec4 object_space_pos = vec4( in_Position.x, in_Position.y, in_Position.z, 1.0);
	 gl_Position = gm_Matrices[MATRIX_WORLD_VIEW_PROJECTION] * object_space_pos;

	 //Now, let's displace the U based on time's current state. I'm using a very small number here because I'm presuming we're using GML's raw time in microseconds here.
	 float displacedU = texPos.x + (iTime*0.000005f);

	 //The next step keeps U clamped between Sprite's values. In real-world production code, the size of the pixel boundary in the texture atlas would have to be considered here, because that's included in the UV coordinates delivered via sprite_get_uvs().
	 //A simple fix would be to do an offset of the pixel boundary value before we do the final clamp, like so:
	 float U1 = UVs[0] + trimDisplacement;
	 float U2 = UVs[2] - trimDisplacement;

	 displacedU  = clamp(displacedU ,U1,U2);

	 //Final U value is passed to fragment shader.
	 v_TexCoord = vec2(displacedU, texPos.y);

	 //Set color, and we're done.
	 v_Color=in_Colour;
}

### Expected Change

when draw_sprite_part(), draw_sprite_part_ext() and draw_sprite_general() are invoked, UVs are kept clamped to the Sprite's UVs

### Steps To Reproduce

1. Start GameMaker
2. Try any of the code with the values shown above.
4. See the issue

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2023.11.1 (Monthly)

### Which operating system(s) are you seeing the problem on?

Windows 11

### Are you running GameMaker from inside your Steam library?

Yes

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [ ] I have included a small sample project

**Comment by JonnoRobson** [comment]
Closing this as I wasn't able to reproduce any obvious issues with the `sprite_draw_part` function. Using a negative "left" or "top" value has the expected effect and does not cause any visual issues. The "x" and "y" parameters appear to be positioning the quad correctly, without affecting the UVs of the quad.

If there are specific visual issues with `sprite_draw_part` when certain values are used then it would be best to open a new ticket and attach a small example project demonstrating the issues.

**Comment by mistletoe** [comment]
Weird!

I must be doing something strange, but I can't think of what; I ran into
this while building simple UI elements (basically, health-bars) that
needed to be rotated or flipped. See the triangular UI elements in the
corner of the image below. To fix the problem, I resorted to using four
distinct images that don't need to be flipped. Something about flipping /
rotating them causes the issue to show up.

I'll try and reproduce it in a small project and let you know.

[image: image.png]
